### ISSUE TITLE
Mesa 21.3.x resulting in flickering on Intel GPUs

### ISSUE THREAD

**Issue by wahjava** [NixOS/nixpkgs#153377]
### Describe the bug

Mesa 21.3.2 present in `nixos-unstable` is causing [flickering in nheko](https://github.com/Nheko-Reborn/nheko/issues/865), and they point to [upstream bug](https://gitlab.freedesktop.org/mesa/mesa/-/issues/5782), not yet fixed as of now. Reverting to nixos-unstable  to Mesa 21.2.6 updates fixes the issue.

### Steps To Reproduce
0. Make sure you're using Intel GPU
1. Switch nixos channel to `59bfda72480`
2. Install `nheko`
3. Start `nheko`

### Expected behavior

No flickering should happen on the application screen. Mesa bug report has a video demonstrating the issue.

### Additional context
* [upstream bug in Mesa](https://gitlab.freedesktop.org/mesa/mesa/-/issues/5782)

### Notify maintainers

cc: @vcunat @primeos 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
- system: `"x86_64-linux"`
- host os: `Linux 5.10.89, NixOS, 22.05 (Quokka)`
- multi-user?: `yes`
- sandbox: `yes`
- version: `nix-env (Nix) 2.5.1`
- channels(root): `"home-manager, nixos-22.05pre342525.59bfda72480"`
- channels(user): `"home-manager"`
- nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by primeos** [comment]
According to the bug report this seems to affect at least Blender as well. Anyway, we can leave this issue open for a while (in case it helps users to figure this out) but I don't think there's anything actionable for Nixpkgs (it's an upstream bug and we simply have to wait until a fix is released via a new version).

It looks like the patch from https://gitlab.freedesktop.org/mesa/mesa/-/issues/5731#note_1205469 should be submitted formally via a PR so that it might get included in Mesa 21.3.4.

**Comment by nrdxp** [comment]
This is also happening on qutebrowser for me, but only on the wayland backend. I have a temporary fix to just launch with `MESA_LOADER_DRIVER_OVERRIDE=llvmpipe`, but this obviously disables the hardware. Works fine though so far for just browsing and work stuff, at least until a proper fix is merged.

**Comment by McSinyx** [comment]
It's unlikely that the patch will be applied for 21.3.4 (as no one filed a MR for it so far), should we apply it downstream?

**Comment by jb55** [comment]
I'm seeing this in qutebrowser as well

**Comment by vcunat** [comment]
I'd prefer to avoid applying a patch that hasn't been reviewed upstream.

**Comment by primeos** [comment]
+1 for what @vcunat said. Luckily a developer from Intel started looking into this so let's see how things will develop: https://gitlab.freedesktop.org/mesa/mesa/-/issues/5731#note_1208981

Ideally a patch will be ready for 21.3.4 but the release might happen in two days (unless it gets delayed) so it'll indeed be close (but we'll of course re-evaluate our options once the release is out).

**Comment by primeos** [comment]
Unfortunately it was indeed too close for the patch to land in Mesa 21.3.4.

However, now there's at least an alternative patch from Paulo Zanoni (the Intel developer) and Giuseppe Ghibò (author of the other patch) and others confirmed that it's working: https://gitlab.freedesktop.org/mesa/mesa/-/issues/5731#note_1214181

I've included that patch in https://github.com/NixOS/nixpkgs/pull/154926. Does someone have time to test it (https://nixos.wiki/wiki/OpenGL#Testing_Mesa_updates) and report if it indeed fixes the issues?

**Comment by magnetophon** [comment]
@primeos I'm happy to test.
Could you confirm I understood the instructions correctly?

1) integrate the code from the above PR into my nixpkgs tree.
2) put the following into my nixos config:
``` nix
system.replaceRuntimeDependencies = [
  ({ original = pkgs.mesa; replacement = (import /srv/nixpkgs-mesa { }).pkgs.mesa; })
  ({ original = pkgs.mesa.drivers; replacement = (import /srv/nixpkgs-mesa { }).pkgs.mesa.drivers; })
];
```
3) replace ``/srv/nixpkgs-mesa`` with the path to my nixpkgs tree.
4) run ``nixos-rebuild test``

Did I get that right?

**Comment by primeos** [comment]
> integrate the code from the above PR into my nixpkgs tree.

It's the easiest if you use a separate/dedicated Nixpkgs tree for Mesa (`/srv/nixpkgs-mesa` in the example). That way you don't have to rebuild a lot of other packages (which would most likely still take way too long). You can also cherry-pick the patches to `nixos-unstable` (I'd recommend doing that) which saves additional rebuilds.

The rest looks fine. And thanks in advance :)

**Comment by phenax** [comment]
> I've included that patch in #154926. Does someone have time to test it (https://nixos.wiki/wiki/OpenGL#Testing_Mesa_updates) and report if it indeed fixes the issues?

It works like a charm. Thanks @primeos!

**Comment by primeos** [comment]
Seems like this wasn't closed automatically once https://github.com/NixOS/nixpkgs/pull/154926 landed in master. Should be fixed now. And upstream merged the required patches in the meantime (s. note in https://github.com/NixOS/nixpkgs/pull/160494).

And thanks for testing the fix @magnetophon and @phenax! :)
